File: src/backend/app/schemas/quote.py
"""
Quote page schemas.

Unified quote data models for the multi-source quote display page.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Data-source metadata
# ---------------------------------------------------------------------------

class DataSourceInfo(BaseModel):
    """Metadata for a single data source (CTP / IB / MT5 / …)."""

    source: str = Field(..., description="Backend identifier, e.g. CTP, IB_WEB, MT5")
    source_label: str = Field(..., description="Display label, e.g. IB")
    status: str = Field(
        ...,
        description=(
            "One of: available, not_configured, not_connected, unavailable"
        ),
    )
    status_message: str | None = Field(None, description="Human-readable status hint")
    capabilities: list[str] = Field(
        default_factory=list,
        description="Supported capabilities, e.g. ['quote', 'search', 'chart']",
    )


class DataSourceListResponse(BaseModel):
    """Response for listing all data sources with their status."""

    sources: list[DataSourceInfo]


# ---------------------------------------------------------------------------
# Quote tick row
# ---------------------------------------------------------------------------

class QuoteTick(BaseModel):
    """A single instrument quote row for the table."""

    source: str = Field(..., description="Backend data-source id")
    source_label: str = Field(..., description="Display label")
    symbol: str = Field(..., description="Instrument code")
    name: str = Field("", description="Instrument name")
    exchange: str = Field("", description="Exchange or category")
    category: str = Field("", description="Instrument category")
    last_price: float | None = Field(None, description="Latest price")
    change: float | None = Field(None, description="Price change amount")
    change_pct: float | None = Field(None, description="Price change percent")
    bid_price: float | None = Field(None, description="Best bid")
    ask_price: float | None = Field(None, description="Best ask")
    high_price: float | None = Field(None, description="Highest price")
    low_price: float | None = Field(None, description="Lowest price")
    open_price: float | None = Field(None, description="Opening price")
    prev_close: float | None = Field(None, description="Previous close")
    volume: float | None = Field(None, description="Volume")
    turnover: float | None = Field(None, description="Turnover / amount")
    open_interest: float | None = Field(None, description="Open interest")
    update_time: str | None = Field(None, description="Last update ISO timestamp")
    status: str = Field("normal", description="Row status: normal, missing, error")
    error_message: str | None = Field(None)


class QuoteListResponse(BaseModel):
    """Batch quote response."""

    source: str
    source_label: str
    total: int
    ticks: list[QuoteTick]
    update_time: str | None = Field(None, description="Server-side fetch timestamp")
    refresh_mode: str = Field(
        "polling",
        description="Current refresh mode: push | polling | manual",
    )


# ---------------------------------------------------------------------------
# Symbol management
# ---------------------------------------------------------------------------

class SymbolItem(BaseModel):
    """A symbol entry (for search results or default list)."""

    symbol: str
    name: str = ""
    exchange: str = ""
    category: str = ""


class SymbolSearchRequest(BaseModel):
    """Search symbols within a data source."""

    source: str = Field(..., description="Backend data-source id")
    keyword: str = Field(..., min_length=1, description="Search keyword")


class SymbolSearchResponse(BaseModel):
    """Search result."""

    source: str
    keyword: str
    results: list[SymbolItem]


class CustomSymbolsRequest(BaseModel):
    """Add or remove custom symbols."""

    source: str = Field(..., description="Backend data-source id")
    symbols: list[str] = Field(..., min_length=1)


class CustomSymbolsResponse(BaseModel):
    """Current custom symbol list after mutation."""

    source: str
    symbols: list[str]


class DefaultSymbolsResponse(BaseModel):
    """Default + custom symbols for a data source."""

    source: str
    default_symbols: list[SymbolItem]
    custom_symbols: list[str]


# ---------------------------------------------------------------------------
# Chart data (P1)
# ---------------------------------------------------------------------------

class KlineBar(BaseModel):
    """A single OHLCV bar for chart rendering."""

    date: str = Field(..., description="Bar datetime string")
    open: float
    high: float
    low: float
    close: float
    volume: float = 0.0


class ChartDataResponse(BaseModel):
    """K-line / chart data response."""

    source: str
    symbol: str
    timeframe: str = Field("M1", description="Timeframe, e.g. M1, M5, M15, H1, D1")
    bars: list[KlineBar]
    total: int = 0

File: src/frontend/src/api/quote.ts
/**
 * 行情报价 API
 */
import request from './index'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface DataSourceInfo {
  source: string
  source_label: string
  status: 'available' | 'not_configured' | 'not_connected' | 'unavailable'
  status_message: string | null
  capabilities: string[]
}

export interface DataSourceListResponse {
  sources: DataSourceInfo[]
}

export interface QuoteTick {
  source: string
  source_label: string
  symbol: string
  name: string
  exchange: string
  category: string
  last_price: number | null
  change: number | null
  change_pct: number | null
  bid_price: number | null
  ask_price: number | null
  high_price: number | null
  low_price: number | null
  open_price: number | null
  prev_close: number | null
  volume: number | null
  turnover: number | null
  open_interest: number | null
  update_time: string | null
  status: string
  error_message: string | null
}

export interface QuoteListResponse {
  source: string
  source_label: string
  total: number
  ticks: QuoteTick[]
  update_time: string | null
  refresh_mode: string
}

export interface SymbolItem {
  symbol: string
  name: string
  exchange: string
  category: string
}

export interface SymbolSearchResponse {
  source: string
  keyword: string
  results: SymbolItem[]
}

export interface DefaultSymbolsResponse {
  source: string
  default_symbols: SymbolItem[]
  custom_symbols: string[]
}

export interface CustomSymbolsResponse {
  source: string
  symbols: string[]
}

export interface KlineBar {
  date: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface ChartDataResponse {
  source: string
  symbol: string
  timeframe: string
  bars: KlineBar[]
  total: number
}

// ---------------------------------------------------------------------------
// API
// ---------------------------------------------------------------------------

export const quoteApi = {
  /** List all data sources with status */
  listSources(): Promise<DataSourceListResponse> {
    return request.get('/quote/sources')
  },

  /** Get default + custom symbols for a data source */
  getSymbols(source: string): Promise<DefaultSymbolsResponse> {
    return request.get('/quote/symbols', { params: { source } })
  },

  /** Add custom symbols */
  addSymbols(source: string, symbols: string[]): Promise<CustomSymbolsResponse> {
    return request.post('/quote/symbols/add', { source, symbols })
  },

  /** Remove custom symbols */
  removeSymbols(source: string, symbols: string[]): Promise<CustomSymbolsResponse> {
    return request.post('/quote/symbols/remove', { source, symbols })
  },

  /** Search symbols within a data source */
  searchSymbols(source: string, keyword: string): Promise<SymbolSearchResponse> {
    return request.get('/quote/symbols/search', { params: { source, keyword } })
  },

  /** Fetch quotes for a data source */
  getQuotes(source: string, symbols?: string[]): Promise<QuoteListResponse> {
    const params: Record<string, string> = { source }
    if (symbols && symbols.length > 0) {
      params.symbols = symbols.join(',')
    }
    return request.get('/quote/ticks', { params })
  },

  /** Fetch K-line / chart data for a symbol */
  getChartData(
    source: string,
    symbol: string,
    timeframe: string = 'M1',
    count: number = 200,
  ): Promise<ChartDataResponse> {
    return request.get('/quote/chart', {
      params: { source, symbol, timeframe, count },
    })
  },
}
